hat/core/src/main/java/hat/codebuilders/C99HATKernelBuilder.java

```diff
@@ -533,6 +515,65 @@ public T hatMemoryLoadOp(ScopedCodeBuilderContext builderContext, HATMemoryLoadO
                 );
     }
 
+    public T hatPtrLoadOp(ScopedCodeBuilderContext builderContext, HATPtrLoadOp hatPtrLoadOp) {
+        ptrAccess(builderContext, hatPtrLoadOp);
+        return self();
+    }
+
+    @Override
+    public T hatPtrStoreOp(ScopedCodeBuilderContext builderContext, HATPtrStoreOp hatPtrStoreOp) {
+        ptrAccess(builderContext, hatPtrStoreOp).equals().recurse(builderContext, ((Op.Result) hatPtrStoreOp.operands().getLast()).op());
+        return self();
+    }
+
+    @Override
+    public T hatPtrLengthOp(ScopedCodeBuilderContext builderContext, HATPtrLengthOp hatPtrLengthOp) {
+        ptrAccess(builderContext, hatPtrLengthOp);
+        return self();
+    }
+
+    T ptrAccess(ScopedCodeBuilderContext builderContext, HATPtrOp hatPtrOp) {
+        boolean isLocalOrPrivateDS = false;
+        if (((Op.Result) hatPtrOp.operands().getFirst()).op() instanceof CoreOp.VarAccessOp.VarLoadOp varLoadOp) {
+            Op resolve = builderContext.scope.resolve(varLoadOp.operands().getFirst());
+            if (resolve instanceof HATMemoryOp) {
+                isLocalOrPrivateDS = true;
+            }
+        }
+        identifier(hatPtrName(hatPtrOp));
+        either(isLocalOrPrivateDS, CodeBuilder::dot, CodeBuilder::rarrow);
+
+        if (hatPtrOp instanceof HATPtrLengthOp) {
+            identifier("length");
+        } else {
+            boolean finalIsLocalOrPrivateDS = isLocalOrPrivateDS;
+            identifier("array").sbrace(_ -> {
+                paren(_ -> identifier("long"));
+                paren(_ -> {
+                    if (hatPtrOp.strides().size() > 1) {
+                        paren(_ -> recurse(builderContext, ((Op.Result) hatPtrOp.operands().get(2)).op()));
+                        asterisk().identifier(hatPtrName(hatPtrOp));
+                        either(finalIsLocalOrPrivateDS, CodeBuilder::dot, CodeBuilder::rarrow).identifier(hatPtrOp.strides() != null ? hatPtrOp.strides().getFirst() : "width");
+                        add().paren(_ -> recurse(builderContext, ((Op.Result) hatPtrOp.operands().get(1)).op()));
+                    } else {
+                        recurse(builderContext, ((Op.Result) hatPtrOp.operands().get(1)).op());
+                    }
+                });
+            });
+        }
+        return self();
+    }
+
+    public String hatPtrName(HATPtrOp hatPtrOp) {
+        Op op = ((Op.Result) ((Op.Result) (hatPtrOp.operands().getFirst())).op().operands().getFirst()).op();
+        return switch (op) {
+            case CoreOp.VarOp varOp -> varOp.varName();
+            case HATLocalVarOp hatLocalVarOp -> hatLocalVarOp.varName();
```

Comment:
Consider using the base class for both Local and Private `HATMemoryOp` and call the varName method. So the switch will be simplified.